Implemented a more reliable custom diff patch processor.

djex opened the pull request:
I have coded a custom diff patch processor as a more reliable alternative to using git apply or wiggle. I have tested the processor and believe I have worked out all the bugs (hopefully - I'm now crossed eye from looking at raw diff patches for a week :D). The processor works very well in my testing. As long as the AI model can generate a somewhat correct diff patch this should be able to apply it. 

The key differences in this new method vs git apply are:
- Majority of the diff header is ignored except file paths.
- Line counts are ignored. However hunk @@ markers are used to determine the start of new hunks.
- White space, line endings and case are ignored in both code and patch.
- The patch is split up into and treated as blocks of search and replace chunks.

At a bare minimum the AI will need to generate a patch with the following:
- Correct +++ and --- file paths
- Correct +, - and ' ' in front of patch lines
- Correct placement of @@ for the start of a new hunk

I have solely been testing with Google Gemini 2.5 Pro and it is working quite well.

robertpiosik commented:
Cool! Will look at it later this week. One thing I have in mind - our current implementation with correct diff is 100% accurate, incorrect diffs when handled with our fallback mechanism of calling refactoring tool are also applied correctly.

I'm more inclined towards api fallback without much thinking if it is really necessary than leading to situation of broken files with programmatic approach.

What do you think?

djex commented:
I agree when using the current method of `git apply` when the diff is 100% correct works very well as it should. However when the diff is not generated correctly containing issues I outlined in my discussion post then `git apply` will fail. I understand your approach with using another AI model to refactor and "figure out" the broken diff has also proven to work well without issue. This is fine as it works.

The reason I coded a custom diff processor which removes many of the common variables that I find AI models get wrong was to prevent reliance on another AI model. I feel it's counter intuitive to use an AI model to work something out that another AI got wrong. Where do we draw the line if that secondary model gets it wrong too?

There is also the concern of cost. Sure there are freely available models now that we can and are taking advantage of but they won't be free forever or will be heavily limited eventually. Potentially having to pay for the main model and then for another model to fix the main models mistakes feels bad in my opinion.

Your implementation now does work well. I have used Gemini 2.5 Pro for main and Gemini 2.0 Flash for refactor when a diff wouldn't apply correctly. I only had one instance where refactor mode using Gemini 2.0 Flash could not apply a malformed diff.

Also seeing that I was the one who initially pushed you to implement a diff mode I also feel somewhat responsible to have it work as best as it can with what we have to work with currently. I foresee eventually the models will advance to the point where this conversation won't need to be had and will produce perfect code and diffs first time. However we are not quite there yet.

Totally up to you if you want to merge the code. Give it a look and see what you think. Maybe implement it as an alternative for those who do not want to use another AI model API for applying a failed diff? Ultimately I felt the problem was simple enough to implement programmatically and could also possibly work in other applications as well so I went for it.

robertpiosik commented:
My initial thoughts are that it should be a fallback mechanism for failed `git apply` — from what I see we skip the native method entirely now.

The perfect flow would be:

1. Try `git apply`
2. For malformed diffs, try to correct them
3. As we're not sure they're applied correctly, always offer option `Looks off, use intelligent mode` for just these corrected by the "patch processor".

I appreciate your work on this, thank you! 🤓

djex commented:
That is a reasonable suggested flow. I will make the changes if I have time later today.

From what I can see the latest pastebin in #236 is suffering from incorrect line counts as expected which the alternative patch processor ignores and should apply correctly if used as a fallback as suggested.

djex commented:
I added back the git apply and implemented the new diff processor as a fall back. I kept the outer try catch as is since at some point I will add .rej file creation to the new diff processor. It will basically create a .rej for any search and replace chunks it can not match 100% similar to what git does.

I will leave it up to you to adjust the error reporting and user prompting to your preference.

robertpiosik commented:
I lack understanding about diffs, can you please explain the purpose behind leaving `.rej` files in directories? Thank you for your work on this 🫶

djex commented:
> I lack understanding about diffs, can you please explain the purpose behind leaving `.rej` files in directories? Thank you for your work on this 🫶

Hey Robert. My apologies on the delay. The weekend got busy and I didn't have much time. The .rej files contain any patches or hunks of a patch that `git apply` could not apply properly. This allows the user to review the .rej files and apply them manually if needed. It's also a good indicator that a particular patch did not apply correctly.

I want to eventually add something similar to the new diff processor which would create .rej files for any search / replace blocks it could not find 100%. This will help integrate better with the current error handling as we can look for any new .rej files in the workspace and either pass them along to the refactoring api or warn the user.

Also I see you completed the merge. Thanks! Also apologies for the mixed formatting. My brain is wired on camel case. I tried to stick to snake case but as you could see that didn't last long. 😄 I will convert it over next time I do a pull request.

robertpiosik commented:
I modified it tiny bit to process patches with deletions only. It's quite reliable! As a false positive can always sneak in, user can still use intelligent update and I think we have nailed it! Cheers
